{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/users/isakova/.local/lib/python3.12/site-packages/anndata/utils.py:429: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/users/isakova/.local/lib/python3.12/site-packages/anndata/utils.py:429: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/users/isakova/.local/lib/python3.12/site-packages/anndata/utils.py:429: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/users/isakova/.local/lib/python3.12/site-packages/anndata/utils.py:429: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/users/isakova/.local/lib/python3.12/site-packages/anndata/utils.py:429: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/users/isakova/.local/lib/python3.12/site-packages/anndata/utils.py:429: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/users/isakova/.local/lib/python3.12/site-packages/anndata/utils.py:429: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/home/users/isakova/.local/lib/python3.12/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/home/users/isakova/.local/lib/python3.12/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/home/users/isakova/.local/lib/python3.12/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/home/users/isakova/.local/lib/python3.12/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/home/users/isakova/.local/lib/python3.12/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(27961, 46173)\n",
      "DDX11L2            ENSG00000290825\n",
      "MIR1302-2HG        ENSG00000243485\n",
      "FAM138A            ENSG00000237613\n",
      "ENSG00000290826    ENSG00000290826\n",
      "OR4F5              ENSG00000186092\n",
      "                        ...       \n",
      "U1                 ENSG00000275987\n",
      "U6                 ENSG00000277890\n",
      "U6                 ENSG00000277927\n",
      "U6                 ENSG00000278625\n",
      "U1                 ENSG00000277374\n",
      "Name: gene_ids, Length: 46173, dtype: object\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/users/isakova/.local/lib/python3.12/site-packages/anndata/_core/anndata.py:1758: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import anndata\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Load your two AnnData objects\n",
    "adata1 = sc.read_10x_h5(\"cellrangerX2024/SSTX-HEK293T-2XDASH_trimmed/outs/filtered_feature_bc_matrix.h5\")\n",
    "# Load the second matrix (unfiltered smallRNA matrix)\n",
    "adata2 = sc.read_10x_h5(\"cellrangerX2025-smallRNA/SSTX-HEK293T-2XDASH_trimmed/outs/raw_feature_bc_matrix.h5\")\n",
    "# Filter adata2 to retain only cell barcodes present in adata1\n",
    "common_barcodes = adata1.obs.index.intersection(adata2.obs.index)\n",
    "adata2 = adata2[common_barcodes, :]\n",
    "\n",
    "# Identify genes in adata2\n",
    "genes_to_remove = set(adata2.var_names)\n",
    "\n",
    "# Filter adata1 to exclude genes present in adata2\n",
    "adata1_filtered = adata1[:, ~adata1.var_names.isin(genes_to_remove)].copy()\n",
    "\n",
    "# Merge datasets based on common gene_ids\n",
    "adata = anndata.concat([adata1_filtered, adata2], axis=1, join='inner')\n",
    "\n",
    "# Check the result\n",
    "print(adata.shape)  # Should be (5, 1) if only 'gene2' is common\n",
    "print(adata.var['gene_ids'])  # Should print ['gene2']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "      <th>GeneName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DDX11L2</th>\n",
       "      <td>ENSG00000290825</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>DDX11L2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM138A</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>FAM138A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000290826</th>\n",
       "      <td>ENSG00000290826</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENSG00000290826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR4F5</th>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U1</th>\n",
       "      <td>ENSG00000275987</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-smallRNA</td>\n",
       "      <td>U1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U6</th>\n",
       "      <td>ENSG00000277890</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-smallRNA</td>\n",
       "      <td>U6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U6</th>\n",
       "      <td>ENSG00000277927</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-smallRNA</td>\n",
       "      <td>U6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U6</th>\n",
       "      <td>ENSG00000278625</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-smallRNA</td>\n",
       "      <td>U6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U1</th>\n",
       "      <td>ENSG00000277374</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-smallRNA</td>\n",
       "      <td>U1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46173 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        gene_ids    feature_types           genome  \\\n",
       "DDX11L2          ENSG00000290825  Gene Expression           GRCh38   \n",
       "MIR1302-2HG      ENSG00000243485  Gene Expression           GRCh38   \n",
       "FAM138A          ENSG00000237613  Gene Expression           GRCh38   \n",
       "ENSG00000290826  ENSG00000290826  Gene Expression           GRCh38   \n",
       "OR4F5            ENSG00000186092  Gene Expression           GRCh38   \n",
       "...                          ...              ...              ...   \n",
       "U1               ENSG00000275987  Gene Expression  GRCh38-smallRNA   \n",
       "U6               ENSG00000277890  Gene Expression  GRCh38-smallRNA   \n",
       "U6               ENSG00000277927  Gene Expression  GRCh38-smallRNA   \n",
       "U6               ENSG00000278625  Gene Expression  GRCh38-smallRNA   \n",
       "U1               ENSG00000277374  Gene Expression  GRCh38-smallRNA   \n",
       "\n",
       "                        GeneName  \n",
       "DDX11L2                  DDX11L2  \n",
       "MIR1302-2HG          MIR1302-2HG  \n",
       "FAM138A                  FAM138A  \n",
       "ENSG00000290826  ENSG00000290826  \n",
       "OR4F5                      OR4F5  \n",
       "...                          ...  \n",
       "U1                            U1  \n",
       "U6                            U6  \n",
       "U6                            U6  \n",
       "U6                            U6  \n",
       "U1                            U1  \n",
       "\n",
       "[46173 rows x 4 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove 'GRCh38_' prefix from the index and create a new column GeneName\n",
    "adata.var['GeneName'] = adata.var.index\n",
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "      <th>GeneName</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>biotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000290825</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>DDX11L2</td>\n",
       "      <td>ENSG00000290825</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000290826</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>ENSG00000290826</td>\n",
       "      <td>ENSG00000290826</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46168</th>\n",
       "      <td>ENSG00000275987</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-smallRNA</td>\n",
       "      <td>U1</td>\n",
       "      <td>ENSG00000275987</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46169</th>\n",
       "      <td>ENSG00000277890</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-smallRNA</td>\n",
       "      <td>U6</td>\n",
       "      <td>ENSG00000277890</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46170</th>\n",
       "      <td>ENSG00000277927</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-smallRNA</td>\n",
       "      <td>U6</td>\n",
       "      <td>ENSG00000277927</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46171</th>\n",
       "      <td>ENSG00000278625</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-smallRNA</td>\n",
       "      <td>U6</td>\n",
       "      <td>ENSG00000278625</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46172</th>\n",
       "      <td>ENSG00000277374</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38-smallRNA</td>\n",
       "      <td>U1</td>\n",
       "      <td>ENSG00000277374</td>\n",
       "      <td>snRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>46173 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_ids    feature_types           genome         GeneName  \\\n",
       "0      ENSG00000290825  Gene Expression           GRCh38          DDX11L2   \n",
       "1      ENSG00000243485  Gene Expression           GRCh38      MIR1302-2HG   \n",
       "2      ENSG00000237613  Gene Expression           GRCh38          FAM138A   \n",
       "3      ENSG00000290826  Gene Expression           GRCh38  ENSG00000290826   \n",
       "4      ENSG00000186092  Gene Expression           GRCh38            OR4F5   \n",
       "...                ...              ...              ...              ...   \n",
       "46168  ENSG00000275987  Gene Expression  GRCh38-smallRNA               U1   \n",
       "46169  ENSG00000277890  Gene Expression  GRCh38-smallRNA               U6   \n",
       "46170  ENSG00000277927  Gene Expression  GRCh38-smallRNA               U6   \n",
       "46171  ENSG00000278625  Gene Expression  GRCh38-smallRNA               U6   \n",
       "46172  ENSG00000277374  Gene Expression  GRCh38-smallRNA               U1   \n",
       "\n",
       "               gene_id         biotype  \n",
       "0      ENSG00000290825          lncRNA  \n",
       "1      ENSG00000243485          lncRNA  \n",
       "2      ENSG00000237613          lncRNA  \n",
       "3      ENSG00000290826          lncRNA  \n",
       "4      ENSG00000186092  protein_coding  \n",
       "...                ...             ...  \n",
       "46168  ENSG00000275987           snRNA  \n",
       "46169  ENSG00000277890           snRNA  \n",
       "46170  ENSG00000277927           snRNA  \n",
       "46171  ENSG00000278625           snRNA  \n",
       "46172  ENSG00000277374           snRNA  \n",
       "\n",
       "[46173 rows x 6 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Specify the path to the GTF file\n",
    "gtf_file = \"/oak/stanford/projects/kibr/Alina/Analysis/F1.method_comparison/filtered_gene_id_biotypes.csv\"\n",
    "\n",
    "# Read the CSV file into a DataFrame\n",
    "gtf_data = pd.read_csv(gtf_file)\n",
    "\n",
    "gtf_data = gtf_data.drop_duplicates(subset='gene_id', keep='first')\n",
    "\n",
    "# Function to append biotype to adata\n",
    "def append_biotype(adata, gtf_data):\n",
    "    # Ensure compatibility of gene_ids column\n",
    "    adata.var = adata.var.copy()  # Avoid potential warnings about chained assignments\n",
    "    adata.var[\"gene_ids\"] = adata.var[\"gene_ids\"].astype(str)\n",
    "    gtf_data[\"gene_id\"] = gtf_data[\"gene_id\"].astype(str)\n",
    "    \n",
    "    # Merge on gene_ids and gene_id\n",
    "    merged_var = adata.var.merge(\n",
    "        gtf_data[[\"gene_id\", \"biotype\"]],  # Only include relevant columns\n",
    "        left_on=\"gene_ids\",                # Match adata.var['gene_ids'] with gtf_data['gene_id']\n",
    "        right_on=\"gene_id\",\n",
    "        how=\"left\"\n",
    "    )\n",
    "    \n",
    "    # Add biotype back to adata.var\n",
    "    adata.var = merged_var\n",
    "    \n",
    "    # Fill missing biotype entries with \"unknown\"\n",
    "    adata.var[\"biotype\"] = adata.var[\"biotype\"].fillna(\"unknown\")\n",
    "    \n",
    "    return adata\n",
    "\n",
    "\n",
    "# Apply the function to each AnnData object in the list\n",
    "adata = append_biotype(adata, gtf_data)\n",
    "\n",
    "adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#adata.var.index = adata.var[\"GeneName\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(\"SSTX_counts_adata_2XDASH.h5ad\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
